[cloud] Added git support to fabric

matt9j opened the pull request:
Summary: Allow fabric to fall back to git if a mercurial repo is not found or mercurial is not installed. This is needed for a real deployment of the git based public projected. As with the previous pull request, please let us know what the standards are for community pull requests if there is anything we should update or change!
Cheers,
-mj

Commit:
[cloud] Added git support to fabric

Summary: Allow fabric to fall back to git if a mercurial repo is not
found or mercurial is not installed.

9muir commented:
Cool, nice addition. I'd probably refactor it a little, see inline comments.

9muir commented on cloud/fabfile.py:
```diff
@@ -46,9 +46,29 @@ def proxy():
+    except:
+        return False
 
 def get_bookmark():
```
I'd make a version of this function that gets all the metadata used in the `package()` function - bookmark/branch (feel free to also make the terminology consistent throughout), version, commit summary - and returns it as a tuple or dict, so that all the git vs hg stuff can be kept in one place. Then you don't need the new `is_foo` functions. `get_bookmark()` would just call that new function and return the bookmark only, for compatibility.

Note that I think function names in the top-level fabfile can (should?) be prefixed with an underscore if they are for internal use only and don't need to be made available as commands. `get_bookmark`/`get_metadata` probably fit into that category, unless they can usefully be run directly for debugging (but whether that justifies keeping them 'exported' is debatable); `is_foo` should be made internal only (if you aren't going to get rid of it/them).

matt9j committed:
Add docstrings to is_* methods in fabfile

Summary: Add appropriate documentation for the provided functions.

matt9j committed:
Make is_* fab functions private

Summary: The is_git and is_hg functions are private helper functions
and shouldn't be exported by fab as callable commands. Use the prefix
underscore convention in Python to denote them private.

matt9j committed:
Factor all fabfile vcs interaction to helper function

Summary: The fabfile extracts some information from version control to
set the deb package version and other package metadata. Instead of
checking for the version control type and doing different actions in
several places, this commit consolidates the vcs interaction to a
single get_versioning_metadata function.

matt9j committed:
Use git rev-parse for finding current branch

Summary: Git provides a `git rev-parse --abbrev-ref` command for
finding the current branch name directly. This is clearer to
understand than just looping through the human readable designed
branch listing. The previous implementation would fail if a branch
name had a `*` in the name. If the git repository is not on a branch
(in a detached head state) this will also not just fall back to
"master", but will instead return "HEAD" for the branch name.

matt9j commented:
@9muir Sorry it took so long for me to get back to this-- it was a busy last quarter :sweat_smile: I believe I took your comments into account and updated the original patch from Kurtis. Let me know if there's anything else you see that I should update!

matt9j committed:
Update other users of get_bookmark

Summary: Update all get_bookmark callsites to the new
_get_versioning_metadata command.

9muir commented on cloud/fabfile.py:
```diff
@@ -47,20 +47,50 @@ def proxy():
+            if "*" in line:
+                 branch = line.split()[1]
+                 break
+    elif(_is_git()):
```
nit: format as `elif _is_git():`.

9muir commented on cloud/fabfile.py:
```diff
@@ -47,20 +47,50 @@ def proxy():
+
+def _get_versioning_metadata():
+    """ Extracts version metadata from the version control system """
+    if (_is_hg()):
```
nit: no parens needed around `_is_hg()`

9muir commented on cloud/fabfile.py:
```diff
@@ -47,20 +47,50 @@ def proxy():
+        branch = local("git rev-parse --abbrev-ref HEAD", capture=True)
+        commit_summary = local('git rev-parse HEAD', capture=True).translate(None, "+")
+    else:
+        raise(Exception("Not git or hg"))
```
nit: no parens required after `raise`

9muir commented on cloud/fabfile.py:
```diff
@@ -215,7 +245,7 @@ def get_machines(environment=None):
 def deploy(description=None):
     """ [deploy] Make a deployment to an environment. """
-    branch = get_bookmark()
+    branch, summary, version = _get_versioning_metadata()
```
by convention you might use underscore as a placeholder for the tuple values you don't care about in this particular case: `branch, _, _ = _get_versioning_metadata()`. Same comment applies below too.

9muir reviewed (changes requested):
Looks good @matt9j, just some minor stylistic nits.

matt9j commented:
> Looks good @matt9j, just some minor stylistic nits.

Thanks for the style tips-- I'm still getting back into writing professional python from C++. I'll address them in a single style update commit!

matt9j committed:
Update to match google style guide and best practices

Summary: This commit addresses a few stylistic nits to better conform
to python best practices and follow the project style guide. This
commit should not change operational behavior.

9muir commented:
Thanks @matt9j, looks good. Will accept to our internal repo, which will get synced and close this out.

matt9j commented:
> Thanks @matt9j, looks good. Will accept to our internal repo, which will get synced and close this out.

:+1:
